### PR TITLE
Add field types

### PR DESCRIPTION
**Pull request by guyyoo**

### CODE REVIEW

File: dataFields/ABFieldTextFormulaCore.js [hunk 1]
```diff
@@ -548,29 +556,29 @@ function getNumberToWords(number) {
...
 function getDateDayOfWeekName(abFactory, date) {
-   // var localizeDT = moment(date);
+   // const localizeDT = moment(date);
    // localizeDT.locale(AD.lang.currentLanguage);
    // return localizeDT.format("dddd");
 
    return abFactory.toDateFormat(date, {
       format: "dddd",
-      localeCode: AD.lang.currentLanguage,
+      localeCode: abFactory.Config._config.user.languageCode,
```
Comment (Hiro-Nakamura) [inline dataFields/ABFieldTextFormulaCore.js:565]: this should be `abFactory.Account.language()`

Better yet, rename `abFactory`  to `AB` to maintain consistency with the rest of our code.

File: dataFields/ABFieldTextFormulaCore.js [hunk 2]
```diff
@@ -548,29 +556,29 @@ function getNumberToWords(number) {
...
 function getDateMonthOfYearName(abFactory, date) {
-   // var localizeDT = moment(date);
+   // const localizeDT = moment(date);
    // localizeDT.locale(AD.lang.currentLanguage);
    // return localizeDT.format("MMMM");
 
    return abFactory.toDateFormat(date, {
       format: "MMMM",
-      localeCode: AD.lang.currentLanguage,
+      localeCode: abFactory.Config._config.user.languageCode,
```
Comment (Hiro-Nakamura) [inline dataFields/ABFieldTextFormulaCore.js:576]: `AB.Account.language()`

Comment (Hiro-Nakamura) [reply dataFields/ABFieldTextFormulaCore.js:576]: just to point out ...

Anytime you reference a private inner value of an object, that is a `no-no`.  For example: `Config._config`  the `_` prefix is supposed to warn you that this is private.  You should be using an exposed method to access any of that data.

**Review by Hiro-Nakamura — CHANGES_REQUESTED** [review]
Some Comments:

Each of our `ABFieldXXXCore` files need some reworking.
- The `ABFieldXXXDefaults` values all need to have their values commented.  Take a look at the `ABFieldStringCore` and `ABFieldNumberCore` examples.  Keep the values in the same order with the same documentation in each of the new `ABFieldXXX` fields you move over to the ABDesigner
- Notice how not every `ABFieldXXXCore` has all the values in their `ABFieldStringDefaults` settings. Take a look at `ABFieldAutoIndexCore` and you'll see that it only defines about 4 of the possible fields.  Go through each of these `ABFieldXXXCore` definitions and fill out all the values.
- in the `ABFieldXXXDefaults`  both the `description`  and `menuName` values should now simply be the `L(key)` that gets displayed.  So remove the current `L()` for those settings and just put the `key`
- anytime you see the old `"Some String #key#".replace("#key#", value);`  replace them with the more modern ``` `Some String ${value}` ``` form.

**Comment by guyyoo** [comment]
@Hiro-Nakamura Thank you so much. I will fix it soon.

File: dataFields/ABFieldConnectCore.js [hunk 3]
```diff
@@ -168,16 +168,16 @@ module.exports = class ABFieldConnectCore extends ABFieldSelectivity {
    relationName() {
       // there is object name - {objectName}.{columnName}
       if (this.columnName.indexOf(".") > -1) {
-         let names = this.columnName.split(".");
+         const names = this.columnName.split(".");
 
          return (
             names[0] +
             "." +
-            (String(names[1]).replace(/[^a-z0-9\.]/gi, "") + "__relation")
```
[the remark this answers is not in the document]
Comment (guyyoo) [reply dataFields/ABFieldConnectCore.js:176]: For this, Prettier tell me the `\` is "Unnecessary escape character: [.eslintno-useless-escape". I am not sure I should ignore?

File: dataFields/ABFieldEmailCore.js [hunk 4]
```diff
@@ -95,9 +96,9 @@ module.exports = class ABFieldEmailCore extends ABField {
     */
    isValidData(data, validator) {
       if (data[this.columnName]) {
-         var Reg = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
```
[the remark this answers is not in the document]
Comment (guyyoo) [reply dataFields/ABFieldEmailCore.js:98]: For this, Prettier tell me the `\` is "Unnecessary escape character: [.eslintno-useless-escape". I am not sure I should ignore?

**Review by Hiro-Nakamura — CHANGES_REQUESTED** [review]
OK, just a couple more things to fix.

File: dataFields/ABFieldEmailCore.js [hunk 5]
```diff
@@ -5,40 +5,69 @@
...
+
+   menuName: "Email",
+   // menuName: what gets displayed in the Editor drop list
+   // NOTE: this will be displayed using a Label: L(menuName)
 
    supportRequire: true,
+   // {bool}
+   // does this ABField support the Required setting?
+
+   supportUnique: false,
```
[the remark this answers is not in the document]
Comment (guyyoo) [reply dataFields/ABFieldEmailCore.js:48]: Ok, I will bring this up 👍

[the remark this answers is not in the document]
Comment (guyyoo) [reply dataFields/ABFieldEmailCore.js:48]: I have talked with the team. I think it should support unique.

**Review by Hiro-Nakamura — APPROVED** [review]
Looking good!

**Pull request merged**